### PR TITLE
[WIP] Switch to .netcoreapp3.0

### PR DESCRIPTION
**Pull request by NicolasDorier**
This build but crashes at runtime because the following dependencies are still stuck in .netcoreapp2.2:

* [x] [Pomelo.EntityFrameworkCore.MySql](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql) (see [following issue](https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/839))
* [ ] [openiddict-core ](https://github.com/openiddict/openiddict-core)
* [ ] [BundlerMinifier PR](https://github.com/madskristensen/BundlerMinifier/pull/444)

Both library have nightly build for 3.0 support, but we should probably better wait for official release.

### CODE REVIEW

**Comment by PinpointTownes** [comment]
FYI, the latest OpenIddict 3.0 alpha packages (3.0.0-alpha1.19504.68) are now compatible with ASP.NET Core 2.1, 2.2 and 3.0, which should muchly simplify your code base, since you'll no longer need to reference different OpenIddict versions to support both ASP.NET Core 2.x and 3.x: https://github.com/openiddict/openiddict-core/issues/736#issuecomment-538601538

File: BTCPayServer/Authentication/OpenId3/ClientCredentialsGrantTypeEventHandler.cs [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+                return;
+            }
+
+            var application = await _applicationManager.FindByClientIdAsync(request.ClientId);
+            if (application == null)
+            {
+                context.Reject(
+                    error: OpenIddictConstants.Errors.InvalidClient,
+                    description: "The client application was not found in the database.");
+                context.HandleRequest();
```
Comment (PinpointTownes) [inline BTCPayServer/Authentication/OpenId3/ClientCredentialsGrantTypeEventHandler.cs]: `context.HandleRequest()` is a way to tell OpenIddict that you want to take care of returning the response yourself (e.g by returning a custom HTTP response). It's likely not what you want to do here. In general, `context.Reject()`, `context.HandleRequest()` and `context.SkipRequest()` are all mutually exclusive.

**Comment by NicolasDorier** [comment]
hey @PinpointTownes I am happy to know OpenIddict3.0 is compatible with 2.1.

Not because we intend to use OpenIddict3.0 for long in 2.1, but because it means we can gradually move the code to 3.0 instead of having one big PR.

The main pain points in the migration have been the following:

* The event handlers do not have access to the HttpContext anymore, so we used the `IHttpContextAccessor` (see BTCPayServer/Authentication/OpenId3/LogoutEventHandler.cs). Still unsure if it is good.
* In our code, somewhere we [were creating an AuthenticationTicket](https://github.com/btcpayserver/btcpayserver/pull/1070/files#diff-16b21d27631a1db3d3202d86e4f7568aL125). Since the scopes are not attached to the ticket anymore, I am unsure how to migrate this code.
* The `options.UseJsonWebTokens();` from openiddict 2.0 disappeared. I don't know how to keep the behavior.

**Comment by PinpointTownes** [comment]
> The event handlers do not have access to the HttpContext anymore, so we used the IHttpContextAccessor (see BTCPayServer/Authentication/OpenId3/LogoutEventHandler.cs). Still unsure if it is good.

As you probably figured out, it's a deliberate change, to make `OpenIddict.Server` completely host-agnostic. What you have is fine, but you can access the ASP.NET Core context by using `context.Transaction.GetHttpRequest().HttpContext` (which doesn't use `IHttpContextAccessor` internally, so no need to register it).

> In our code, somewhere we were creating an AuthenticationTicket. Since the scopes are not attached to the ticket anymore, I am unsure how to migrate this code.

All the extensions have been moved to `ClaimsPrincipal`, as `AuthenticationTicket` is an ASP.NET Core type we couldn't depend on.

> The options.UseJsonWebTokens(); from openiddict 2.0 disappeared. I don't know how to keep the behavior.

JWT is now the default token format for all types of tokens (authorization codes and refresh tokens are always encrypted using JWE, so don't try to read them using jwt.io 😄).

**Comment by NicolasDorier** [comment]
Thanks for your help, @PinpointTownes, another confusing part: Should we use `OpenIddictServerAspNetCoreDefaults.AuthenticationScheme` or `OpenIdConnectDefaults.AuthenticationScheme` as authentication scheme?

**Comment by NicolasDorier** [comment]
@PinpointTownes I must say, I am a bit afraid about our use of OpenIddict. I don't understand enough the design, and I fear there is huge security hole in our code as a result.

Right now we don't use OpenIddict for anything important, but we will. Can I pay for your time later on so you can make a review of our OpenIddict code once I finished completely the 3.0 migration? If yes, contact me on nicolas.dorier@gmail.com

**Comment by PinpointTownes** [comment]
> Should we use OpenIddictServerAspNetCoreDefaults.AuthenticationScheme or OpenIdConnectDefaults.AuthenticationScheme as authentication scheme?

`OpenIddictServerAspNetCoreDefaults.AuthenticationScheme` when you want to return an OIDC authorization or token response (with a token).

`OpenIdConnectDefaults.AuthenticationScheme` when you want to trigger an OIDC authentication dance from an OIDC ASP.NET Core client.

**Comment by NicolasDorier** [comment]
We still need https://github.com/madskristensen/BundlerMinifier/pull/443 

I am unsure if I will migrate to 3.0 since 3.1 is LTS and should release next week. But at least I brought the gap to make .NET 3.0 work.

**Comment by NicolasDorier** [comment]
I officially give up on .netcoreapp3.0

I keep this PR open just to have something to experiment to keep making the gap to cross smaller between netcore2.1.

Will reconsider on netcoreapp3.1

**Comment by NicolasDorier** [comment]
There is still a bug in open iddict https://github.com/openiddict/openiddict-core/issues/834 that make most of the test fail. Bug I think there is also other bug not caught by tests related to MVC.

**Comment by bolatovumar** [comment]
@NicolasDorier just wondering, have you abandoned this PR for now?

**Comment by NicolasDorier** [comment]
@bolatovumar yeah, was waiting for .net 3.1, which is now out.
There was some bugs in OpenIddict which prevented me to continue, but those should now be fixed, and if not fixed,  I would just stop supporting it until it is fixed (we don't use it yet anyway)

**Comment by PinpointTownes** [comment]
@NicolasDorier just to be clear, the "bug" mentioned earlier is not an OpenIddict regression, it's an EF Core breaking change (it no longer generates IDs for string primary keys by default). And it's certainly not blocking as you can fix it on your side by assigning the ID in the constructor (e.g https://github.com/openiddict/openiddict-core/blob/dev/src/OpenIddict.EntityFrameworkCore.Models/OpenIddictAuthorization.cs#L18-L22)

https://github.com/openiddict/openiddict-core/issues/834 tracks a possible future improvement consisting in calling `.ValueGeneratedOnAdd()` for you.

**Comment by NicolasDorier** [comment]
@PinpointTownes I think the problematic classes are in OpenIddict assembly, so if you did it on your side, I should just be able to update assembly and it would work right?

**Comment by PinpointTownes** [comment]
They are not. OpenIddict exposes the base generic types, but you own the final entities: https://github.com/btcpayserver/btcpayserver/blob/master/BTCPayServer.Data/Data/BTCPayOpenIdAuthorization.cs

It’s worth noting this "issue" is not specific to OpenIddict: you’d have the same issue with Identity if you inherited from their generic entities.

**Comment by Kukks** [comment]
But you should be marking them, on the base types, inside OpenIddict?

**Comment by PinpointTownes** [comment]
@Kukks marking? Not sure what you mean exactly.

**Pull request closed without merging**